### PR TITLE
Don't require semicolons before unary operators

### PR DESCRIPTION
**Pull request by adrianheine**
**What is the purpose of this pull request? (put an "X" next to item)**

[ ] Documentation update
[] Bug fix ([template](https://github.com/eslint/eslint/blob/master/templates/bug-report.md))
[ ] New rule ([template](https://github.com/eslint/eslint/blob/master/templates/rule-proposal.md))
[X] Changes an existing rule ([template](https://github.com/eslint/eslint/blob/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

**What rule do you want to change?**

semi

**Does this change cause the rule to produce more or fewer warnings?**

more (with `"never"`)

**How will the change be implemented? (New option, new default behavior, etc.)?**

Change to behavior of `"never"` option

**Please provide some example code that this change will affect:**

```js
a;
++b;
```

**What does the rule currently do for this code?**

Give only one error.

**What will the rule do after it's changed?**

Give two errors.

**Please check each item to ensure your pull request is ready:**

- [ ] I've read the [pull request guide](http://eslint.org/docs/developer-guide/contributing/pull-requests)
- [X] I've included tests for my change
- [ ] I've updated documentation for my change (if appropriate)

**What changes did you make? (Give an overview)**

**Is there anything you'd like reviewers to focus on?**

### CODE REVIEW

**Comment by platinumazure** [comment]
> **Does this change cause the rule to produce more or fewer warnings?**
> 
> **fewer (with `"never"`)**

(snip)

> **What does the rule currently do for this code?**
> 
> **Give only one error.**
> 
> **What will the rule do after it's changed?**
> 
> **Give two errors.**

Those sections of the PR template conflict with each other. Please clarify.

**Comment by adrianheine** [comment]
Sorry, it's obviously more errors.

**Comment by not-an-aardvark** [comment]
We should verify that the semicolon there doesn't change semantics at all due. I think it's fine in this case due to ASI:

```js
// this increments bar
foo
++bar

// this increments foo
foo++
bar

// this increments bar
foo
++
bar
```

However, could you add a test for this case? It should not report any errors.

```js
/* eslint semi: [2, "always"] */
++
foo
```

**Comment by adrianheine** [comment]
@not-an-aardvark Added the test you requested.

File: lib/rules/semi.js [hunk 1]
```diff
@@ -53,7 +53,7 @@ module.exports = {
 
     create(context) {
 
-        const OPT_OUT_PATTERN = /[\[\(\/\+\-]/; // One of [(/+-
+        const OPT_OUT_PATTERN = /[\[\(\/](\+[^\+]|\-[^\-])/; // One of [(/+-, but not ++ or --
```
Comment (mysticatea) [inline lib/rules/semi.js]: Could you add tests like below?

    var a = b;
    + c

This semicolon cannot be removed.
But this fix seems to make the rule warning this semicolon.

---

I think this RegExp can be more simple.

    const OPT_OUT_PATTERN = /^[[(/+-]$/;

Comment (adrianheine) [inline lib/rules/semi.js]: Thanks, I simplified the regular expression, fixed it for your example and added it as a test.

Comment (mysticatea) [inline lib/rules/semi.js]: It looks complex to me. :(
Is just adding `^` and `$` not enough?

Comment (adrianheine) [inline lib/rules/semi.js]: Seems to work fine for me.

**Comment by mysticatea** [comment]
I changed label to `bug` since I think this is the bug to overlook unnecessary semicolons.

**Pull request merged**